xlators/storage/posix/src/posix-inode-fd-ops.c

```diff
@@ -2791,20 +2791,21 @@ posix_setxattr(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *dict,
                 goto unlock;
             }
 
+            /* using 4k to fit in PATH_MAX in one of the below operation */
             cs_var = alloca(4096);
-            sprintf(cs_var, "%" PRId64, tmp_stbuf.ia_mtime);
-
-            /*TODO: may be should consider nano-second also */
-            if (strncmp(cs_var, tdata->data, tdata->len) > 0) {
+            len = sprintf(cs_var, "%" PRId64 ".%09" PRIu32, tmp_stbuf.ia_mtime,
+                          tmp_stbuf.ia_mtime_nsec);
+            if (strncmp(cs_var, user_mtime, len)) {
                 gf_msg(this->name, GF_LOG_ERROR, 0, 0,
-                       "mtime "
-                       "passed is different from seen by file now."
-                       " Will skip truncating the file");
+                       "mtime passed is different from seen by file now. Will "
+                       "skip truncating the file (%s on file vs %s passed)",
+                       cs_var, user_mtime);
                 ret = -1;
                 op_errno = EINVAL;
                 goto unlock;
             }
 
+            memset(cs_var, 0, 4096);
```

Comment:
I'm not sure this (and other changes above) are related to this patch. I also don't understand why this memset is needed.